[Java16] Test Failure: com.ibm.ws.request.timing.hung.fat.HungRequestTiming.testSequentialHungMultipleRequests

RTC 283107 / 283170

Issue caused by Java 16 - Hotspot, where it can not create thread dumps for hung requests. Need to permit illegal access of classes in the test case, when using anything above Java 8.

```
testSequentialHungMultipleRequests:junit.framework.AssertionFailedError: 2021-03-06-05:25:00:251 Expected 1 Java core.. But found : 0
at com.ibm.ws.request.timing.hung.fat.HungRequestTiming.testSequentialHungMultipleRequests(HungRequestTiming.java:434)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:197)
at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:318)
at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:171)
at componenttest.rules.repeater.RepeatTests$CompositeRepeatTestActionStatement.evaluate(RepeatTests.java:115)
```

```
>Stack Dump = java.lang.IllegalStateException: java.lang.IllegalAccessException: class com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine cannot access class sun.tools.attach.HotSpotVirtualMachine (in module jdk.attach) because module jdk.attach does not export sun.tools.attach to unnamed module @7dc5e7b4
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine.remoteDataDump(HotSpotJavaDumperImpl.java:341)
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl.createThreadDump(HotSpotJavaDumperImpl.java:195)
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl.dump(HotSpotJavaDumperImpl.java:101)
> at com.ibm.ws.kernel.launch.internal.FrameworkManager.dumpJava(FrameworkManager.java:1233)
> at com.ibm.ws.kernel.launch.internal.FrameworkManager.dumpJava(FrameworkManager.java:1219)
> at com.ibm.ws.kernel.launch.internal.LibertyProcessImpl.createJavaDump(LibertyProcessImpl.java:106)
> at com.ibm.ws.request.timing.manager.ThreadDumpManager$1.run(ThreadDumpManager.java:108)
> at com.ibm.ws.threading.internal.SchedulingRunnableFixedHelper.run(SchedulingRunnableFixedHelper.java:290)
> at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
> at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
> at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
> at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
> at java.base/java.lang.Thread.run(Thread.java:831)
>Caused by: java.lang.IllegalAccessException: class com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine cannot access class sun.tools.attach.HotSpotVirtualMachine (in module jdk.attach) because module jdk.attach does not export sun.tools.attach to unnamed module @7dc5e7b4
> at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:385)
> at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:687)
> at com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine.remoteDataDump(HotSpotJavaDumperImpl.java:339)
```

--- a/dev/com.ibm.ws.request.timing.hung_fat/fat/src/com/ibm/ws/request/timing/hung/fat/HungRequestEnableThreadDumps.java
+++ b/dev/com.ibm.ws.request.timing.hung_fat/fat/src/com/ibm/ws/request/timing/hung/fat/HungRequestEnableThreadDumps.java
@@ -34,11 +34,12 @@ import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.JavaInfo;
 import componenttest.topology.impl.LibertyServer;
 
 /**
  *
- * This test class contains test cases that test the Hung Request Timing server configuration attribute enableThreadDumps, when enabl
+ * This test class contains test cases that test the Hung Request Timing server configuration attribute enableThreadDumps.
  *
  */
 @RunWith(FATRunner.class)
@@ -52,8 +53,14 @@ public class HungRequestEnableThreadDumps {
 
     @BeforeClass
     public static void setUp() throws Exception {
+        JavaInfo java = JavaInfo.forCurrentVM();
         ShrinkHelper.defaultDropinApp(server, "TestWebApp", "com.ibm.testwebapp");
-        CommonTasks.writeLogMsg(Level.INFO, " Starting server..");
+        int javaVersion = java.majorVersion();
+        if (javaVersion != 8) {
+            CommonTasks.writeLogMsg(Level.INFO, " Java version = " + javaVersion + " - It is higher than 8, adding --add-exports...");
+            server.copyFileToLibertyServerRoot("add-exports/jvm.options");
+        }
+        CommonTasks.writeLogMsg(Level.INFO, " Starting server...");
         server.startServer();
     }
 

--- a/dev/com.ibm.ws.request.timing.hung_fat/fat/src/com/ibm/ws/request/timing/hung/fat/HungRequestTiming.java
+++ b/dev/com.ibm.ws.request.timing.hung_fat/fat/src/com/ibm/ws/request/timing/hung/fat/HungRequestTiming.java
@@ -47,6 +47,7 @@ import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.JavaInfo;
 import componenttest.topology.impl.LibertyServer;
 
 @RunWith(FATRunner.class)
@@ -64,8 +65,14 @@ public class HungRequestTiming {
 
     @BeforeClass
     public static void setUp() throws Exception {
+        JavaInfo java = JavaInfo.forCurrentVM();
         ShrinkHelper.defaultDropinApp(server, "TestWebApp", "com.ibm.testwebapp");
-        CommonTasks.writeLogMsg(Level.INFO, " starting server..");
+        int javaVersion = java.majorVersion();
+        if (javaVersion != 8) {
+            CommonTasks.writeLogMsg(Level.INFO, " Java version = " + javaVersion + " - It is higher than 8, adding --add-exports...");
+            server.copyFileToLibertyServerRoot("add-exports/jvm.options");
+        }
+        CommonTasks.writeLogMsg(Level.INFO, " starting server...");
         server.startServer();
     }
 

--- a/dev/com.ibm.ws.request.timing.hung_fat/publish/files/add-exports/jvm.options
+++ b/dev/com.ibm.ws.request.timing.hung_fat/publish/files/add-exports/jvm.options
@@ -1,0 +1,3 @@
+# for dump action: com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine.remoteDataDump(HotSpotJavaDumperImpl.java:342)
+--add-exports
+jdk.attach/sun.tools.attach=ALL-UNNAMED

--- a/dev/com.ibm.ws.request.timing_fat/fat/src/com/ibm/ws/request/timing/fat/TimingRequestTiming.java
+++ b/dev/com.ibm.ws.request.timing_fat/fat/src/com/ibm/ws/request/timing/fat/TimingRequestTiming.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2014, 2020 IBM Corporation and others.
+ * Copyright (c) 2014, 2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -39,6 +39,7 @@ import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.JavaInfo;
 import componenttest.topology.impl.LibertyServer;
 
 @RunWith(FATRunner.class)
@@ -60,8 +61,14 @@ public class TimingRequestTiming {
 
     @BeforeClass
     public static void setUp() throws Exception {
+        JavaInfo java = JavaInfo.forCurrentVM();
         ShrinkHelper.defaultDropinApp(server, "jdbcTestPrj_3", "com.ibm.ws.request.timing");
-        CommonTasks.writeLogMsg(Level.INFO, " starting server..");
+        int javaVersion = java.majorVersion();
+        if (javaVersion != 8) {
+            CommonTasks.writeLogMsg(Level.INFO, " Java version = " + javaVersion + " - It is higher than 8, adding --add-exports...");
+            server.copyFileToLibertyServerRoot("add-exports/jvm.options");
+        }
+        CommonTasks.writeLogMsg(Level.INFO, " starting server...");
         server.startServer();
     }
 

--- a/dev/com.ibm.ws.request.timing_fat/publish/files/add-exports/jvm.options
+++ b/dev/com.ibm.ws.request.timing_fat/publish/files/add-exports/jvm.options
@@ -1,0 +1,3 @@
+# for dump action: com.ibm.ws.kernel.boot.internal.commands.HotSpotJavaDumperImpl$VirtualMachine.remoteDataDump(HotSpotJavaDumperImpl.java:342)
+--add-exports
+jdk.attach/sun.tools.attach=ALL-UNNAMED

